File: backend/utils/sanitize.py
"""
HTML sanitize 工具（D4 决策：XSS 防护）

为避免 XSS 攻击，对所有富文本输入（评论 content、task description 等）做清洗。
- 白名单标签：常用排版标签
- 白名单属性：a 标签的 href/title/target
- 白名单协议：http、https、mailto
- 剥除不在白名单的标签和属性（保留文本内容）
- 剥除 HTML 注释
"""
import bleach

ALLOWED_TAGS = [
    'p', 'br', 'strong', 'em', 'u', 'a',
    'ul', 'ol', 'li', 'blockquote', 'code', 'pre',
    'h1', 'h2', 'h3',
]
ALLOWED_ATTRS = {
    'a': ['href', 'title', 'target'],
}
ALLOWED_PROTOCOLS = ['http', 'https', 'mailto']


def sanitize_html(content):
    """清洗 HTML，返回安全的子集

    Args:
        content: 原始 HTML 字符串

    Returns:
        清洗后的 HTML 字符串。如果 content 为 None 或空字符串，原样返回。
    """
    if not content:
        return content
    return bleach.clean(
        content,
        tags=ALLOWED_TAGS,
        attributes=ALLOWED_ATTRS,
        protocols=ALLOWED_PROTOCOLS,
        strip=True,           # 不允许的标签被剥除（保留文本）
        strip_comments=True,  # 剥除 HTML 注释
    )


def strip_tags(content):
    """剥除所有 HTML 标签，仅保留纯文本

    用于 title 字段：title 不应包含任何 HTML（即使在白名单内的标签），
    因为前端 title 是用 v-text 渲染而非 v-html，HTML 标签会显示为字面字符。

    Args:
        content: 原始字符串

    Returns:
        仅含纯文本的字符串。如果 content 为 None 或空字符串，原样返回。
    """
    if not content:
        return content
    return bleach.clean(
        content,
        tags=[],
        attributes={},
        strip=True,
    ).strip()

File: frontend/src/utils/sanitize.js
// B0291/B0292 — 前端 XSS 防御
// 后端已用 bleach sanitize (RR1)，前端再加一道防御性 strip
// 仅允许白名单标签 + 属性，禁止 javascript:/data:/on* 事件

const ALLOWED_TAGS = new Set([
  'p', 'br', 'strong', 'em', 'u', 's', 'code', 'pre', 'blockquote',
  'h1', 'h2', 'h3', 'h4', 'h5', 'h6',
  'ul', 'ol', 'li',
  'a', 'img', 'span', 'div',
])
const ALLOWED_ATTRS = new Set(['href', 'title', 'alt', 'src', 'class', 'target', 'rel'])

// 简单实现：去掉 script/iframe/on* 属性 / javascript: 协议
// 实际项目建议加 dompurify 依赖（~20KB），这里用最简的 regex strip
export function sanitizeHtml(html) {
  if (!html || typeof html !== 'string') return ''
  let s = html
  // 1. 移除 <script>...</script>
  s = s.replace(/<script[\s\S]*?<\/script>/gi, '')
  // 2. 移除 <iframe>, <object>, <embed>
  s = s.replace(/<\/?(iframe|object|embed|form|input|style|link|meta)[\s\S]*?>/gi, '')
  // 3. 移除 on* 事件属性
  s = s.replace(/\son\w+\s*=\s*("[^"]*"|'[^']*'|[^\s>]+)/gi, '')
  // 4. 移除 javascript: / data: 协议
  s = s.replace(/(href|src)\s*=\s*("javascript:[^"]*"|'javascript:[^']*'|javascript:[^\s>]+)/gi, '')
  s = s.replace(/(href|src)\s*=\s*("data:(?!image\/)[^"]*"|'data:(?!image\/)[^']*')/gi, '')
  // 5. 移除未白名单标签（保留内容）
  s = s.replace(/<(\/?)([a-z][a-z0-9]*)([^>]*)>/gi, (m, slash, tag, attrs) => {
    if (!ALLOWED_TAGS.has(tag.toLowerCase())) return ''
    // 过滤属性
    const filtered = (attrs.match(/[a-z\-]+\s*=\s*("[^"]*"|'[^']*')/gi) || [])
      .filter(a => {
        const name = a.split('=')[0].trim().toLowerCase()
        return ALLOWED_ATTRS.has(name)
      })
      .join(' ')
    return `<${slash}${tag}${filtered ? ' ' + filtered : ''}>`
  })
  return s
}
